includes/API/ApiLabkiUpdate.php

```diff
@@ -11,165 +11,336 @@
+        foreach ( $pages as $p ) {
+            $relPath = $p['path'] ?? null;
+            $finalTitle = $p['finalTitle'] ?? '';
+            $sourceName = $p['original'] ?? $p['name'] ?? '';
+            if ( $relPath === null || $finalTitle === '' ) {
+                continue;
+            }
+
+            $fileUrl = rtrim( $repoUrl, '/' ) . '/' . ltrim( $relPath, '/' );
+            $wikitext = $this->fetchWikiFile( $fileUrl );
+            if ( $wikitext === null ) {
+                wfDebugLog( 'Labki', "Failed to fetch wiki file: $fileUrl" );
+                continue;
+            }
+
+            // Rewrite internal links
+            $updatedText = $this->rewriteLinks( $wikitext, $rewriteMap );
+
+            $title = Title::newFromText( $finalTitle );
+            if ( !$title ) {
+                continue;
+            }
+
+            $content = new WikitextContent( $updatedText );
+            $page = WikiPage::factory( $title );
+            $editResult = $page->doEditContent( $content, 'Imported via LabkiPackManager' );
+
+            if ( !$editResult->isOK() ) {
+                wfDebugLog( 'Labki', "Edit failed for $finalTitle" );
+                continue;
+            }
+
+            $pagesOut[] = [
+                'name' => $sourceName,
+                'finalTitle' => $finalTitle,
+                'pageNamespace' => $title->getNamespace(),
+                'wikiPageId' => $page->getId()
+            ];
+        }
+
+        return $pagesOut;
+    }
+
+    private function fetchWikiFile( string $url ): ?string {
+        $ctx = stream_context_create( [ 'http' => [ 'timeout' => 10 ] ] );
+        $data = @file_get_contents( $url, false, $ctx );
+        return $data !== false ? $data : null;
+    }
+
+    private function rewriteLinks( string $text, array $map ): string {
+        foreach ( $map as $orig => $final ) {
+            if ( $orig === $final ) {
+                continue;
+            }
+            $escapedOrig = preg_quote( $orig, '/' );
+
+            // Replace [[Page]] or [[Page|...]]
+            $text = preg_replace(
+                '/\[\[' . $escapedOrig . '(\|[^]]*)?\]\]/u',
+                '[[' . $final . '$1]]',
+                $text
+            );
+
+            // Replace {{Template}} or {{Template|...}}
+            $text = preg_replace(
+                '/\{\{' . $escapedOrig . '(\|[^}]*)?\}\}/u',
+                '{{' . $final . '$1}}',
+                $text
+            );
+        }
+        return $text;
+    }
+
+    private function buildRewriteMap( string $repoUrl, array $incomingPacks ): array {
+        $pageReg = new LabkiPageRegistry();
+        $repoReg = new LabkiRepoRegistry();
+
+        $repoId = $repoReg->ensureRepo( $repoUrl );
+        $map = $pageReg->getRewriteMapForRepo( $repoId );
+
+        foreach ( $incomingPacks as $pack ) {
+            foreach ( $pack['pages'] ?? [] as $pg ) {
+                $orig = str_replace( ' ', '_', $pg['original'] ?? $pg['name'] ?? '' );
```

Comment:
I think we need both here, a wikilink in a page can be made like `[[Some Page]]` and `[[Some_Page]]` and they're equivalent (they can also be made like `[[some Page]]` since first letters are autocapitalized too, but maybe handle main cases then edge cases). 

Is there any way we can use mediawiki page methods for this? It looks like there are accessor methods on the parsed page text, might be good to just directly work with the mediawiki API here because it probably handles all the weird variations in links and whatnot https://www.mediawiki.org/wiki/Manual:ParserOutput.php